staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go

```diff
@@ -330,57 +330,68 @@ func TestPodResourceNameCompletionFuncJointFormTooManyArgs(t *testing.T) {
-	pods, _, _ := cmdtesting.TestData()
-	addResourceToFactory(tf, pods)
-
-	compFunc := PodResourceNameAndContainerCompletionFunc(tf)
-	comps, directive := compFunc(cmd, []string{"pod-name", "container-name"}, "")
-	checkCompletion(t, comps, []string{}, directive, cobra.ShellCompDirectiveNoFileComp)
-}
-
-func TestPodResourceNameAndContainerCompletionFuncJointFormNoArgs(t *testing.T) {
-	tf, cmd := prepareCompletionTest()
-	pods, _, _ := cmdtesting.TestData()
-	addResourceToFactory(tf, pods)
-
-	compFunc := PodResourceNameAndContainerCompletionFunc(tf)
-	comps, directive := compFunc(cmd, []string{}, "pod/b")
-	checkCompletion(t, comps, []string{"pod/bar"}, directive, cobra.ShellCompDirectiveNoFileComp)
-}
-
-func TestPodResourceNameAndContainerCompletionFuncJointFormTooManyArgs(t *testing.T) {
-	tf, cmd := prepareCompletionTest()
-	pods, _, _ := cmdtesting.TestData()
-	addResourceToFactory(tf, pods)
+func TestResourceAndContainerNameCompletionFunc(t *testing.T) {
+	barPod := getTestPod()
+	testCases := []struct {
+		name              string
+		args              []string
+		toComplete        string
+		expectedComps     []string
+		expectedDirective cobra.ShellCompDirective
+	}{
+		{
+			name:              "no args pod name",
+			args:              []string{},
+			toComplete:        "b",
+			expectedComps:     []string{"bar"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+		{
+			name:              "no args resources",
+			args:              []string{},
+			toComplete:        "s",
+			expectedComps:     []string{"services/", "statefulsets/"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace,
+		},
+		{
+			name:              "joint from no args",
+			args:              []string{},
+			toComplete:        "pod/b",
+			expectedComps:     []string{"pod/bar"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+		{
+			name:              "joint from too many args",
+			args:              []string{"pod/pod-name", "container-name"},
+			toComplete:        "",
+			expectedComps:     []string{},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+		{
+			name:              "complete all containers' names",
+			args:              []string{"bar"},
+			toComplete:        "",
+			expectedComps:     []string{"bar", "foo"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+		{
+			name:              "complete specific container name",
+			args:              []string{"bar"},
+			toComplete:        "b",
+			expectedComps:     []string{"bar"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+	}
 
-	compFunc := PodResourceNameAndContainerCompletionFunc(tf)
-	comps, directive := compFunc(cmd, []string{"pod/pod-name", "container-name"}, "")
-	checkCompletion(t, comps, []string{}, directive, cobra.ShellCompDirectiveNoFileComp)
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tf, cmd := prepareCompletionTest()
```

Comment:
These 3 lines are the same for every test. No need to rerun them for each test. Are you able to move these lines outside the for loop?
```
			tf, cmd := prepareCompletionTest()
			addResourceToFactory(tf, barPod)
			compFunc := PodResourceNameAndContainerCompletionFunc(tf)
```

Reply:
Good catch, thanks!

Reply:
OK, I see the unit tests failed after doing this, so I think you should put it back as it was.  I didn't realize it needed to be inside the loop.